File: docs/breez-sdk/snippets/wasm/htlcs.ts
import { createHash } from 'crypto'
import type { Payment, BreezSdk, PrepareSendPaymentResponse } from '@breeztech/breez-sdk-spark'

const exampleSendHtlcPayment = async (sdk: BreezSdk): Promise<Payment> => {
  // ANCHOR: send-htlc-payment
  const paymentRequest = '<spark address>'
  // Set the amount you wish to pay the receiver
  const amountSats = BigInt(50_000)

  const prepareResponse = await sdk.prepareSendPayment({
    paymentRequest,
    amount: amountSats,
    tokenIdentifier: undefined,
    conversionOptions: undefined,
    feePolicy: undefined
  })

  // If the fees are acceptable, continue to create the HTLC Payment
  if (prepareResponse.paymentMethod.type === 'sparkAddress') {
    const fee = prepareResponse.paymentMethod.fee
    console.debug(`Fees: ${fee} sats`)
  }

  const preimage = '<32-byte unique preimage hex>'
  const preimageBuffer = Buffer.from(preimage, 'hex')
  const paymentHash = createHash('sha256').update(preimageBuffer).digest('hex')

  const sendResponse = await sdk.sendPayment({
    prepareResponse,
    options: {
      type: 'sparkAddress',
      htlcOptions: {
        paymentHash,
        expiryDurationSecs: 1000
      }
    }
  })
  const payment = sendResponse.payment
  // ANCHOR_END: send-htlc-payment
  return payment
}

const exampleReceiveHodlInvoicePayment = async (sdk: BreezSdk) => {
  // ANCHOR: receive-hodl-invoice-payment
  const preimage = '<32-byte unique preimage hex>'
  const preimageBuffer = Buffer.from(preimage, 'hex')
  const paymentHash = createHash('sha256').update(preimageBuffer).digest('hex')

  const response = await sdk.receivePayment({
    paymentMethod: {
      type: 'bolt11Invoice',
      description: 'HODL invoice',
      amountSats: 50_000,
      expirySecs: undefined,
      paymentHash
    }
  })

  const invoice = response.paymentRequest
  console.log(`HODL invoice: ${invoice}`)
  // ANCHOR_END: receive-hodl-invoice-payment
}

const exampleListClaimableHtlcPayments = async (sdk: BreezSdk): Promise<Payment[]> => {
  // ANCHOR: list-claimable-htlc-payments
  const response = await sdk.listPayments({
    typeFilter: ['receive'],
    statusFilter: ['pending'],
    paymentDetailsFilter: [{
      type: 'spark',
      htlcStatus: ['waitingForPreimage']
    }, {
      type: 'lightning',
      htlcStatus: ['waitingForPreimage']
    }],
    assetFilter: undefined
  })
  const payments = response.payments

  for (const payment of payments) {
    if (payment.details?.type === 'spark' && payment.details.htlcDetails != null) {
      console.log(`Spark HTLC expiry time: ${payment.details.htlcDetails.expiryTime}`)
    } else if (payment.details?.type === 'lightning') {
      console.log(`Lightning HTLC expiry time: ${payment.details.htlcDetails.expiryTime}`)
    }
  }
  // ANCHOR_END: list-claimable-htlc-payments
  return payments
}

const exampleClaimHtlcPayment = async (sdk: BreezSdk): Promise<Payment> => {
  // ANCHOR: claim-htlc-payment
  const preimage = '<preimage hex>'
  const response = await sdk.claimHtlcPayment({
    preimage
  })
  const payment = response.payment
  // ANCHOR_END: claim-htlc-payment
  return payment
}

File: docs/breez-sdk/snippets/swift/BreezSdkSnippets/Sources/Htlcs.swift
import BigNumber
import BreezSdkSpark
import CryptoKit
import Foundation

extension Data {
    init?(hexString: String) {
        let scanner = Scanner(string: hexString)
        var data = Data()
        while !scanner.isAtEnd {
            var byte: UInt64 = 0
            if scanner.scanHexInt64(&byte) {
                data.append(UInt8(byte))
            } else {
                return nil
            }
        }
        self = data
    }

    func hexEncodedString() -> String {
        return self.map { String(format: "%02hhx", $0) }.joined()
    }
}

func sendHtlcPayment(sdk: BreezSdk) async throws -> Payment {
    // ANCHOR: send-htlc-payment
    let paymentRequest = "<spark address>"
    // Set the amount you wish to pay the receiver
    let amountSats = BInt(50_000)
    let prepareRequest = PrepareSendPaymentRequest(
        paymentRequest: paymentRequest,
        amount: amountSats,
        tokenIdentifier: nil,
        conversionOptions: nil,
        feePolicy: nil
    )
    let prepareResponse = try await sdk.prepareSendPayment(request: prepareRequest)

    // If the fees are acceptable, continue to create the HTLC Payment
    if case let .sparkAddress(_, fee, _) = prepareResponse.paymentMethod {
        print("Fees: \(fee) sats")
    }

    let preimage = "<32-byte unique preimage hex>"
    let preimageData = Data(hexString: preimage)!
    let paymentHashDigest = SHA256.hash(data: preimageData)
    let paymentHash = Data(paymentHashDigest).hexEncodedString()

    // Set the HTLC options
    let htlcOptions = SparkHtlcOptions(
        paymentHash: paymentHash,
        expiryDurationSecs: 1000
    )
    let options = SendPaymentOptions.sparkAddress(htlcOptions: htlcOptions)

    let request = SendPaymentRequest(
        prepareResponse: prepareResponse,
        options: options
    )
    let sendResponse = try await sdk.sendPayment(request: request)
    let payment = sendResponse.payment
    // ANCHOR_END: send-htlc-payment
    return payment
}

func receiveHodlInvoicePayment(sdk: BreezSdk) async throws {
    // ANCHOR: receive-hodl-invoice-payment
    let preimage = "<32-byte unique preimage hex>"
    let preimageData = Data(hexString: preimage)!
    let paymentHashDigest = SHA256.hash(data: preimageData)
    let paymentHash = Data(paymentHashDigest).hexEncodedString()

    let response = try await sdk.receivePayment(
        request: ReceivePaymentRequest(
            paymentMethod: ReceivePaymentMethod.bolt11Invoice(
                description: "HODL invoice",
                amountSats: 50_000,
                expirySecs: nil,
                paymentHash: paymentHash
            )
        )
    )

    let invoice = response.paymentRequest
    print("HODL invoice: \(invoice)")
    // ANCHOR_END: receive-hodl-invoice-payment
}

func listClaimableHtlcPayments(sdk: BreezSdk) async throws -> [Payment] {
    // ANCHOR: list-claimable-htlc-payments
    let request = ListPaymentsRequest(
        typeFilter: [PaymentType.receive],
        statusFilter: [PaymentStatus.pending],
        paymentDetailsFilter: [
            PaymentDetailsFilter.spark(
                htlcStatus: [SparkHtlcStatus.waitingForPreimage],
                conversionRefundNeeded: nil
            ),
            PaymentDetailsFilter.lightning(
                htlcStatus: [SparkHtlcStatus.waitingForPreimage]
            ),
        ]
    )

    let response = try await sdk.listPayments(request: request)
    let payments = response.payments

    for payment in payments {
        if case let .spark(_, htlcDetails, _) = payment.details, let htlc = htlcDetails {
            print("Spark HTLC expiry time: \(htlc.expiryTime)")
        } else if case let .lightning(_, _, _, htlcDetails, _, _, _) = payment.details {
            print("Lightning HTLC expiry time: \(htlcDetails.expiryTime)")
        }
    }
    // ANCHOR_END: list-claimable-htlc-payments
    return payments
}

func claimHtlcPayment(sdk: BreezSdk) async throws -> Payment {
    // ANCHOR: claim-htlc-payment
    let preimage = "<preimage hex>"
    let response = try await sdk.claimHtlcPayment(
        request: ClaimHtlcPaymentRequest(preimage: preimage)
    )
    let payment = response.payment
    // ANCHOR_END: claim-htlc-payment
    return payment
}
